File: webhooks/csharp/Utils/DynamicForkTask.cs
namespace WebhooksCodelab.Utils;

/// <summary>
/// DYNAMIC_FORK task that runs one branch per task definition it receives at
/// runtime. Use it in place of the SDK's <c>DynamicFork</c>, and add a
/// <c>JoinTask</c> immediately after it in the workflow so the workflow waits
/// for every branch to finish.
/// </summary>
/// <example>
/// <code>
/// var fork = new DynamicForkTask(
///     "send_emails_fork",
///     getEmailsTask.Output("dynamicTasks"),
///     getEmailsTask.Output("dynamicTasksInputs"));
/// workflow.WithTask(getEmailsTask, fork, new JoinTask("send_emails_join"), ...);
/// </code>
/// </example>
public class DynamicForkTask : Conductor.Definition.TaskType.Task
{
    // The input parameters that hold the forked task definitions and their inputs.
    private const string TasksParam = "dynamicTasks";
    private const string TasksInputsParam = "dynamicTasksInputs";

    /// <param name="taskReferenceName">Reference name of the fork task.</param>
    /// <param name="tasks">
    /// Expression for the list of task definitions to fork, each with a
    /// <c>name</c>, a unique <c>taskReferenceName</c> and a <c>type</c>
    /// (<c>SIMPLE</c> for a worker task).
    /// </param>
    /// <param name="tasksInputs">
    /// Expression for a map from each forked task's <c>taskReferenceName</c> to
    /// that task's input.
    /// </param>
    public DynamicForkTask(string taskReferenceName, string tasks, string tasksInputs)
        : base(taskReferenceName, WorkflowTaskTypeEnum.FORKJOINDYNAMIC)
    {
        // conductor-csharp 3.0.0's DynamicFork only sets these two fields when
        // it is converted by a code path that ConductorWorkflow.WithTask never
        // calls, so the server would not know where to find the forked tasks.
        DynamicForkTasksParam = TasksParam;
        DynamicForkTasksInputParamName = TasksInputsParam;
        InputParameters[TasksParam] = tasks;
        InputParameters[TasksInputsParam] = tasksInputs;
    }
}

File: webhooks/typescript/utils/dynamicForkTask.ts
import type { WorkflowTask } from "@io-orkes/conductor-javascript";

// The input parameters that hold the forked task definitions and their inputs.
const TASKS_PARAM = "dynamicTasks";
const TASKS_INPUTS_PARAM = "dynamicTasksInputs";

/**
 * DYNAMIC_FORK task that runs one branch per task definition it receives at runtime. Use it in
 * place of the SDK's dynamicForkTask, which only accepts a list of tasks that is fixed when the
 * workflow is built, and add the SDK's joinTask straight after it so the workflow waits for every
 * branch to finish.
 *
 * @example
 * // taskOutput is defined in deployWaitForWebhookWorkflow.ts.
 * const fork = dynamicForkTask(
 *   "send_emails_fork",
 *   taskOutput(getEmailsTask, "dynamicTasks"),
 *   taskOutput(getEmailsTask, "dynamicTasksInputs"),
 * );
 * workflow.add([getEmailsTask, fork, joinTask("send_emails_join", []), ...]);
 *
 * @param taskReferenceName Reference name of the fork task.
 * @param tasks Expression for the list of task definitions to fork, each with a `name`, a unique
 *   `taskReferenceName` and a `type` (`SIMPLE` for a worker task).
 * @param tasksInputs Expression for an object that maps each forked task's `taskReferenceName`
 *   to that task's input.
 */
export function dynamicForkTask(
  taskReferenceName: string,
  tasks: string,
  tasksInputs: string,
): WorkflowTask {
  return {
    name: taskReferenceName,
    taskReferenceName,
    type: "FORK_JOIN_DYNAMIC",
    inputParameters: {
      [TASKS_PARAM]: tasks,
      [TASKS_INPUTS_PARAM]: tasksInputs,
    },
    dynamicForkTasksParam: TASKS_PARAM,
    dynamicForkTasksInputParamName: TASKS_INPUTS_PARAM,
  };
}
